gluten-data/src/main/scala/io/glutenproject/execution/GlutenRowToArrowColumnarExec.scala

```diff
@@ -272,38 +297,119 @@ case class GlutenRowToArrowColumnarExec(child: SparkPlan)
+                arrowBuf.close()
+                arrowBuf = tmpBuf
+              }
+              Platform.copyMemory(unsafeRow.getBaseObject, unsafeRow.getBaseOffset,
+                null, arrowBuf.memoryAddress() + offset, sizeInBytes)
+              offset += sizeInBytes
+              rowLength += sizeInBytes.toLong
+              rowCount += 1
+            }
+            numInputRows += rowCount
+            numOutputBatches += 1
+            val cSchema = ArrowSchema.allocateNew(allocator)
+            val cArray = ArrowArray.allocateNew(allocator)
+            try {
+              GlutenArrowAbiUtil.exportSchema(allocator, arrowSchema, cSchema)
+              jniWrapper.nativeConvertRowToColumnar(
+                cSchema.memoryAddress(), rowLength.toArray,
+                arrowBuf.memoryAddress(), cArray.memoryAddress(),
+                NativeMemoryAllocators.contextInstance().getNativeInstanceId)
+              val cb = GlutenArrowAbiUtil.importToSparkColumnarBatch(allocator, arrowSchema,
+                cArray)
+              val newColumns = (0 until cb.numCols).map(
+                cb.column).toArray
+              new ColumnarBatch(newColumns, cb.numRows())
+            } finally {
+              arrowBuf.close()
+              arrowBuf = null
+              cSchema.close()
+              cArray.close()
+            }
+          }
+
+          def javaConvert(row: InternalRow): ColumnarBatch = {
+            logInfo("Not UnsafeRow, fallback to java based r2c")
             val vectors: Seq[WritableColumnVector] =
               ArrowWritableColumnVector.allocateColumns(numRows, schema)
             var rowCount = 0
+
+            converter.convert(row, vectors.toArray)
+            rowCount += 1
+
             while (rowCount < numRows && rowIterator.hasNext) {
               val row = rowIterator.next()
-              val start = System.nanoTime()
-              converters.convert(row, vectors.toArray)
-              elapse += System.nanoTime() - start
+              converter.convert(row, vectors.toArray)
               rowCount += 1
             }
-            vectors.foreach(v => v.asInstanceOf[ArrowWritableColumnVector].setValueCount(rowCount))
-            convertTime.set(NANOSECONDS.toMillis(elapse))
+            vectors.foreach(v => v.asInstanceOf[ArrowWritableColumnVector]
+              .setValueCount(rowCount))
             numInputRows += rowCount
             numOutputBatches += 1
-            last_cb = new ColumnarBatch(vectors.toArray, rowCount)
-            last_cb
+            new ColumnarBatch(vectors.toArray, rowCount)
+          }
+
+          override def next(): ColumnarBatch = {
+            val firstRow = rowIterator.next()
+            val start = System.currentTimeMillis()
+            val cb = firstRow match {
+              case unsafeRow: UnsafeRow if typeCheckNative() =>
+                nativeConvert(unsafeRow)
+              case _ =>
+                javaConvert(firstRow)
```

Comment:
Using the first row to decide type is usually sub-optimal. I'll open a issue for it too.

If it's not possible to fetch some metadata about the row type, it might be feasible to check all rows belong to one batch to ensure they are of the same type.

Reply:
https://github.com/oap-project/gluten/issues/1121

Reply:
It not uses the firstRow to check the type, I lost to send `localSchema` to check, now it use `plan.schema` to check the type

Reply:
I know schema was passed through JNI but I meant we are using the first row to decide whether the rest of rows are `UnsafeRow`s or not.